lib/model/query/datasets.js

```diff
@@ -128,8 +128,27 @@ const _getByNameSql = ((fields, datasetName, projectId) => sql`
 
 // Creates or merges dataset, properties and field mapping.
 // Expects dataset:Frame auxed with `formDefId` and array of field:Frame auxed with `propertyName`
-const createOrMerge = (dataset, fields) => ({ all }) =>
-  all(_createOrMerge(dataset, fields));
+// Note from Kathleen: I apologize at how hacky it is to get the provisioned actee ID into this
+// and get the created dataset out again...
+const createOrMerge = (dataset, fields) => ({ all, Actees, Datasets, Projects }) =>
+  Promise.all([
+    Projects.getById(dataset.projectId).then((o) => o.get()),
+    Datasets.getByProjectAndName(dataset.projectId, dataset.name)
+  ])
+    .then(([project, datasetOption]) =>
+      (datasetOption.isDefined()
+        ? datasetOption.get().acteeId
+        : Actees.provision('dataset', project).then((actee) => (actee.id))))
+    .then((acteeId) =>
+      all(_createOrMerge(dataset, fields, acteeId)))
+    .then(() => Datasets.getByProjectAndName(dataset.projectId, dataset.name))
+    .then((ds) => ds.get());
+
+createOrMerge.audit = (dataset, _, fields) => (log) =>
+  ((dataset.revisionNumber === 0)
+    ? log('dataset.create', dataset, { fields: fields.map((f) => [f.path, f.propertyName]) })
+    : log('dataset.update', dataset, { fields: fields.map((f) => [f.path, f.propertyName]) }));
```

Comment:
New audit log actions generally need to be added to `actionCondition()` in lib/model/query/audits.js. There's no `if` there for datasets, so I think that'll need to be added. It should be possible to request /v1/audits?action=dataset and have the API return audit log entries for both `dataset.create` and `dataset.update`.